### ISSUE TITLE
Canceling LivingSpawnEvent.SpecialSpawn does not prevent mob spawn

### ISSUE THREAD

**Issue by JeremyJBrennan** [MinecraftForge/MinecraftForge#8404]
**Minecraft Version:** 1.18.1

**Forge Version:** 39.0.44

**Steps to Reproduce:**
 1. Create a LivingSpawnEvent.SpecialSpawn listener method. Have it cancel.

    @SubscribeEvent
    public static void livingSpawnEvent$SpecialSpawnListener(final LivingSpawnEvent.SpecialSpawn event){
        event.setCanceled(true);
    }

Issue:
All spawns that should be prevented by canceling this event still occur.

I originally noticed this while canceling skeleton spawns and noticing they were still spawning, just without bows, presumably because the finalizeSpawn function where the bow is given to skeleton wasn't being reached.

**Comment by sciwhiz12** [comment]
Confirmed in 1.18.1-39.0.66 by creating a test mod which cancels that event for creepers and placing a creeper spawner in the world via `/setblock ~ ~ ~ minecraft:spawner{Delay:20s,MaxNearbyEntities:6s,MaxSpawnDelay:100s,MinSpawnDelay:90s,RequiredPlayerRange:16s,SpawnCount:1s,SpawnPotentials:[{data:{entity:{id:"minecraft:creeper"}},weight:1}],SpawnRange:4s}`. 

Expected behavior (according to the javadocs) is that the creepers are not spawned by the spawner. Actual behavior is that the creepers are being spawned. Compare with using a creeper spawn egg, which results in the expected behavior of not spawning a creeper.

The code which fires the `SpecialSpawn` event (the `ForgeEventFactory#doSpecialSpawn` method) is called in three places: [`EntityType#spawn(ServerLevel, CompoundTag, Component, Player, BlockPos, MobSpawnType, boolean, boolean)`][entitytype], [`BaseSpawner#serverTick`][basespawner], and [`NaturalSpawner#spawnCategoryForPosition(MobCategory, ServerLevel, ChunkAccess, BlockPos, SpawnPredicate, AfterSpawnCallback)`][naturalspawner]. 

The first path handles the event cancellation as expected and prevents the mob from being spawned. However, the latter two paths do not prevent the entity from spawning; they prevent the `Mob#finalizeSpawn` method from being called, but the actual entity is still spawned in the world.

[entitytype]: https://github.com/MinecraftForge/MinecraftForge/blob/3b754fef5e5008259a0cc6f4830140d06078e92c/patches/minecraft/net/minecraft/world/entity/EntityType.java.patch#L50
[basespawner]: https://github.com/MinecraftForge/MinecraftForge/blob/3b754fef5e5008259a0cc6f4830140d06078e92c/patches/minecraft/net/minecraft/world/level/BaseSpawner.java.patch#L13
[naturalspawner]: https://github.com/MinecraftForge/MinecraftForge/blob/3b754fef5e5008259a0cc6f4830140d06078e92c/patches/minecraft/net/minecraft/world/level/NaturalSpawner.java.patch#L27

**Comment by Shadows-of-Fire** [comment]
From a conversation with lex on discord a while ago, cancelling the spawn is not the desired behavior (the javadocs are the ones incorrect) - cancellation of finalizeSpawn is the intended case.

**Comment by Shadows-of-Fire** [comment]
Convo link for reference, (internal channel): https://discord.com/channels/313125603924639766/887479168168775761/942495250298077204
However, given the discrepancy between the callsites, this issue should remain open until something is done.  It might be advisable to change the title though.

**Comment by McJty** [comment]
Seems like EntityType.spawn() IS wrong though:

```
   @Nullable
   public T spawn(ServerLevel pLevel, @Nullable CompoundTag pCompound, @Nullable Component pCustomName, @Nullable Player pPlayer, BlockPos pPos, MobSpawnType pSpawnType, boolean pShouldOffsetY, boolean pShouldOffsetYMore) {
      T t = this.create(pLevel, pCompound, pCustomName, pPlayer, pPos, pSpawnType, pShouldOffsetY, pShouldOffsetYMore);
      if (t != null) {
         if (t instanceof net.minecraft.world.entity.Mob && net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn((net.minecraft.world.entity.Mob) t, pLevel, pPos.getX(), pPos.getY(), pPos.getZ(), null, pSpawnType)) return null;
         pLevel.addFreshEntityWithPassengers(t);
      }

      return t;
   }
```

This code does actually cancel the spawn if the event is canceled. That's different from the behaviour in NaturalSpawner. This means that spawn eggs don't work for example.

**Comment by LexManos** [comment]
Correct, this one should be canEntitySpawn not doSpecialSpawn. doSpecialSpawn is for canceling shit like spider jockeys and other 'special' features of spawning. Not the whole thing.

**Comment by Shadows-of-Fire** [comment]
This is addressed by https://github.com/MinecraftForge/MinecraftForge/pull/9133 but I may want to wire up a more minor PR to fix this call site while we await 19.3
